# HG changeset patch
# User Ben Abbott  <bpabbott@mac.com>
# Date 1330216408 18000
#      Sat Feb 25 19:33:28 2012 -0500
# Node ID 4f160344236adb7e943e5524615451c7db651dd3
# Parent  c4a2eac243b4fd3e45a048d276887150052e0dc8
2011-04-26 Daniel Wagenaar  <daw@caltech.edu>

* graphics.cc: Modified the text::properties::get_extent method
to return correct (x0,y0,w,h). Added FIXME comment to
text::properties::get_extent_matrix method, and general comments
to text::properties::update_text_extent method.

diff --git a/src/graphics.cc b/src/graphics.cc
--- a/src/graphics.cc
+++ b/src/graphics.cc
@@ -6877,23 +6877,31 @@ text::properties::get_data_position (voi
     pos = convert_text_position (pos, *this, get_units (), "data");
 
   return pos;
 }
 
 Matrix
 text::properties::get_extent_matrix (void) const
 {
+  // FIXME: Should this function also add the (x,y) base position?
   return extent.get ().matrix_value ();
 }
 
 octave_value
 text::properties::get_extent (void) const
 {
+  // FIXME: This doesn't work right for 3D plots.
+  // (It doesn't in Matlab either, at least not in version 6.5.)
   Matrix m = extent.get ().matrix_value ();
+  Matrix pos = get_position ().matrix_value ();
+  Matrix p = convert_text_position (pos, *this, get_units (), "pixels");
+
+  m(0) += p(0);
+  m(1) += p(1);
 
   return convert_text_position (m, *this, "pixels", get_units ());
 }
 
 void
 text::properties::update_font (void)
 {
 #ifdef HAVE_FREETYPE
@@ -6931,16 +6939,20 @@ text::properties::update_text_extent (vo
   // FIXME: string should be parsed only when modified, for efficiency
 
   octave_value string_prop = get_string ();
 
   string_vector sv = string_prop.all_strings ();
 
   renderer.text_to_pixels (sv.join ("\n"), pixels, bbox,
                            halign, valign, get_rotation ());
+  /* The bbox is relative to the text's position.
+     We'll leave it that way, because get_position() does not return
+     valid results when the text is first constructed.
+     Conversion to proper coordinates is performed in get_extent. */
   set_extent (bbox);
 
 #endif
 
   if (autopos_tag_is ("xlabel") || autopos_tag_is ("ylabel") ||
       autopos_tag_is ("zlabel") || autopos_tag_is ("title"))
     update_autopos ("sync");
 }
